src/target/linux/writemem.rs

```diff
@@ -0,0 +1,301 @@
+    // Obviously, this is very inefficient since it requires a lot of context switches,
+    // but sometimes it's the only way to overwrite the target's memory.
+
+    // Break write ops into groups of <usize> bytes.
+    let write_op_groups = write_ops.iter().flat_map(|op| op.as_ptrace());
+
+    for op in write_op_groups {
+        assert!(op.source_len <= WORD_SIZE);
+
+        if op.source_len < WORD_SIZE {
+            // Write op is smaller than a single word, so we should read memory before rewriting it.
+            let mut word = ptrace::read(pid, op.remote_base as *mut _)?.to_ne_bytes();
+            let src_bytes: &[u8] = slice::from_raw_parts(op.source_ptr as *const _, op.source_len);
+
+            for offset in 0..op.source_len {
+                word[offset] = src_bytes[offset];
+            }
+
+            ptrace::write(
+                pid,
+                op.remote_base as *mut _,
+                usize::from_ne_bytes(word) as *mut usize as *mut _,
+            )?;
+        } else {
+            let word = op.source_ptr.cast::<usize>().read();
+            ptrace::write(pid, op.remote_base as *mut _, word as *mut _)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Allows to write data to different locations in debuggee's memory as a single operation.
+/// It requires a memory page to be writable.
+pub(crate) unsafe fn write_process_vm(
+    pid: Pid,
+    write_ops: &[WriteOp],
+) -> Result<(), Box<dyn std::error::Error>> {
+    // Create a list of `IoVec`s and remote `IoVec`s
+    let remote_iov = write_ops
+        .iter()
+        .map(WriteOp::as_remote_iovec)
+        .collect::<Vec<_>>();
+
+    let local_iov = write_ops
+        .iter()
+        .map(WriteOp::as_local_iovec)
+        .collect::<Vec<_>>();
+
+    let bytes_written = libc::process_vm_writev(
+        pid.into(),
+        local_iov.as_ptr(),
+        local_iov.len() as libc::c_ulong,
+        remote_iov.as_ptr(),
+        remote_iov.len() as libc::c_ulong,
+        0,
+    );
+
+    if bytes_written == -1 {
+        // fixme: return a proper error type
+        return Err(Box::new(nix::Error::last()));
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{write_process_vm, write_ptrace, WriteMemory, WriteOp};
+    use libc::c_void;
+    use nix::{
+        sys::{ptrace, signal, wait},
+        unistd::{fork, getpid, getppid, ForkResult},
+    };
+    use std::mem;
+
+    #[test]
+    fn write_memory_proc_vm() {
+        let var: usize = 52;
+        let var2: u8 = 128;
+
+        let write_var_op: usize = 0;
+        let write_var2_op: u8 = 0;
+
+        let write_mem = WriteMemory::new(getpid())
+            .write(&var, &write_var_op as *const _ as usize)
+            .write(&var2, &write_var2_op as *const _ as usize);
+
+        unsafe {
+            write_process_vm(write_mem.pid, &write_mem.write_ops).expect("Failed to write memory")
+        };
+
+        assert_eq!(write_var2_op, var2);
+        assert_eq!(write_var_op, var);
```

Comment:
These should use `volatile_read`, otherwise LLVM assumes that nobody writes to them.